crates/runtime/src/datasets_health_monitor.rs

```diff
@@ -0,0 +1,229 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::{
+    collections::HashMap,
+    sync::Arc,
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
+use arrow::array::RecordBatch;
+use datafusion::{
+    datasource::TableProvider, error::DataFusionError, execution::context::SessionContext,
+    sql::TableReference,
+};
+use futures::{future::join_all, stream::TryStreamExt};
+use snafu::{ResultExt, Snafu};
+use tokio::sync::Mutex;
+
+use crate::component::dataset::Dataset;
+
+const DATASETS_AVAILABILITY_CHECK_INTERVAL_SECONDS: u64 = 60; // every minute
+const DATASET_UNAVAILABLE_THRESHOLD_SECONDS: u64 = 10 * 60; // 10 minutes
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Unable to get table: {source}"))]
+    UnableToGetTable { source: DataFusionError },
+}
+
+#[derive(Clone)]
+pub struct DatasetAvailabilityInfo {
+    name: String,
+    table_provider: Arc<dyn TableProvider>,
+    last_available_time: SystemTime,
+}
+
+impl DatasetAvailabilityInfo {
+    pub fn new(name: String, table_provider: Arc<dyn TableProvider>) -> Self {
+        Self {
+            name,
+            table_provider,
+            last_available_time: SystemTime::now(),
+        }
+    }
+}
+
+pub enum AvailabilityVerificationResult {
+    Available,
+    Unavailable(SystemTime, String),
+}
+
+pub struct DatasetsHealthMonitor {
+    df_ctx: Arc<SessionContext>,
+    monitored_datasets: Arc<Mutex<HashMap<String, Arc<DatasetAvailabilityInfo>>>>,
```

Comment:
Isn't it better to use RwLock here?

Reply:
RwLock is better than there are many reads, in this case we mostly write/mutate `monitored_datasets` so Mutex is appropriate here.